File: src/output.zig
const std = @import("std");

pub fn printUsage(writer: anytype) !void {
    try writer.writeAll(
        \\shlice
        \\
        \\  shlice start [--id <shell-id>] -- <custom-command...>
        \\  shlice exec [--id <shell-id>] [--timeout <seconds>] <command>
        \\  echo "<command>" | shlice exec [--id <shell-id>]
        \\  shlice stop [<shell-id>]
        \\  shlice status [--id <shell-id>]
        \\  shlice list
        \\
        \\Help flags: -h, --help, -?
        \\Default shell id: main
        \\
    );
}

pub fn printStarted(id: []const u8) !void {
    try std.io.getStdOut().writer().print("started {s}\n", .{id});
}

pub fn printStopped(id: []const u8) !void {
    try std.io.getStdOut().writer().print("stopped {s}\n", .{id});
}

pub fn printError(message: []const u8) !void {
    try std.io.getStdErr().writer().print("error: {s}\n", .{message});
}

pub fn printStatusHeader(writer: anytype) !void {
    try writer.writeAll("id\tstatus\tpid\tbroker\tcmd\tcwd\n");
}

pub fn printNoShells(writer: anytype) !void {
    try writer.writeAll("no shells\n");
}

pub fn printStatusLine(writer: anytype, id: []const u8, status: []const u8, pid: ?u32, broker_pid: ?u32, command_line: []const u8, cwd: []const u8) !void {
    if (pid) |live_pid| {
        if (broker_pid) |live_broker_pid| {
            try writer.print("{s}\t{s}\t{d}\t{d}\t{s}\t{s}\n", .{ id, status, live_pid, live_broker_pid, command_line, cwd });
        } else {
            try writer.print("{s}\t{s}\t{d}\t-\t{s}\t{s}\n", .{ id, status, live_pid, command_line, cwd });
        }
    } else {
        if (broker_pid) |live_broker_pid| {
            try writer.print("{s}\t{s}\t-\t{d}\t{s}\t{s}\n", .{ id, status, live_broker_pid, command_line, cwd });
        } else {
            try writer.print("{s}\t{s}\t-\t-\t{s}\t{s}\n", .{ id, status, command_line, cwd });
        }
    }
}

test "usage mentions exec and list" {
    var buffer = std.ArrayList(u8).init(std.testing.allocator);
    defer buffer.deinit();
    try printUsage(buffer.writer());
    try std.testing.expect(std.mem.indexOf(u8, buffer.items, "shlice exec") != null);
    try std.testing.expect(std.mem.indexOf(u8, buffer.items, "shlice list") != null);
}

File: src/output.rs
use std::io::{self, Write};

pub fn print_usage(mut writer: impl Write) -> io::Result<()> {
    writer.write_all(
        b"shlice\n\n  shlice start [--id <shell-id>] -- <command...>\n  shlice exec [--id <shell-id>] [--timeout <seconds>] <command>\n  echo \"<command>\" | shlice exec [--id <shell-id>]\n  shlice stop [<shell-id>]\n  shlice status [--id <shell-id>]\n  shlice list\n\nHelp flags: -h, --help, -?\nDefault shell id: main\n",
    )
}

pub fn print_started(id: &str) {
    let _ = writeln!(io::stdout(), "started {id}");
}

pub fn print_stopped(id: &str) {
    let _ = writeln!(io::stdout(), "stopped {id}");
}

pub fn print_error(msg: &str) {
    let _ = writeln!(io::stderr(), "error: {msg}");
}

pub fn print_status_header(mut writer: impl Write) -> io::Result<()> {
    writer.write_all(b"id\tstatus\tpid\tbroker\tcmd\n")
}

pub fn print_no_shells(mut writer: impl Write) -> io::Result<()> {
    writer.write_all(b"no shells\n")
}

pub fn print_status_line(
    mut writer: impl Write,
    id: &str,
    status: &str,
    pid: Option<u32>,
    broker_pid: Option<u32>,
    command_line: &str,
) -> io::Result<()> {
    match (pid, broker_pid) {
        (Some(pid), Some(broker_pid)) => writeln!(
            writer,
            "{id}\t{status}\t{pid}\t{broker_pid}\t{command_line}"
        ),
        (Some(pid), None) => writeln!(writer, "{id}\t{status}\t{pid}\t-\t{command_line}"),
        (None, Some(broker_pid)) => {
            writeln!(writer, "{id}\t{status}\t-\t{broker_pid}\t{command_line}")
        }
        (None, None) => writeln!(writer, "{id}\t{status}\t-\t-\t{command_line}"),
    }
}
